### PR TITLE
Python Test Submission

### PR DESCRIPTION
**Pull request by meistermuka**

### CODE REVIEW

File: alayatodo/views.py [hunk 1]
```diff
@@ -44,39 +61,115 @@ def logout():
...
+        abort(404)
+
     return render_template('todo.html', todo=todo)
 
 
+@app.route('/todo/<id>/json', methods=['GET'])
+@todo_login
+def todo_json(id):
+
+    todo = Todos.query.filter_by(id=id).first()
```
Comment (m-a-choquette) [inline alayatodo/views.py]: When filtering on the primary key, you can use the `get` method.

But here this allows a user to see a todo from another user.

File: alayatodo/views.py [hunk 2]
```diff
@@ -44,39 +61,115 @@ def logout():
...
+
+    error = None
+
+    description = request.form.get('description')
+
+    if not description:
+        error = 'Description required!'
+
+    if error is not None:
+        flash(error)
```
Comment (m-a-choquette) [inline alayatodo/views.py]: This validation check seems a little bit verbose/complicated.

File: alayatodo/views.py [hunk 3]
```diff
@@ -44,39 +61,115 @@ def logout():
...
+    else:
+        user_id = session['user']['id']
+        description = request.form.get('description', '')
```
Comment (m-a-choquette) [inline alayatodo/views.py]: You already have the `description` variable holding the same value.

File: alayatodo/views.py [hunk 4]
```diff
@@ -44,39 +61,115 @@ def logout():
...
 
 @app.route('/todo/<id>', methods=['POST'])
+@todo_login
 def todo_delete(id):
-    if not session.get('logged_in'):
-        return redirect('/login')
-    g.db.execute("DELETE FROM todos WHERE id ='%s'" % id)
-    g.db.commit()
+
+    todo = Todos.query.filter_by(id=id).first()
```
Comment (m-a-choquette) [inline alayatodo/views.py]: This allows a user to delete a todo from another user.

File: alayatodo/views.py [hunk 5]
```diff
@@ -44,39 +61,115 @@ def logout():
...
+    flash("'%s' description has been removed!" % description)
+
     return redirect('/todo')
+
+
+@app.route('/complete/<id>', methods=['POST'])
+@todo_login
+def todo_complete(id):
+
+    todo = Todos.query.filter_by(id=id).first()
```
Comment (m-a-choquette) [inline alayatodo/views.py]: Allows a user to complete a todo from another user.

File: alayatodo/views.py [hunk 6]
```diff
@@ -44,39 +61,115 @@ def logout():
...
+    todo.completed = not bool(todo.completed)
```
Comment (m-a-choquette) [inline alayatodo/views.py]: This will raise an `AttributeError` if the todo id doesn't exist.

File: alayatodo/views.py [hunk 7]
```diff
@@ -44,39 +61,115 @@ def logout():
...
+@todo_login
 def todo_delete(id):
-    if not session.get('logged_in'):
-        return redirect('/login')
-    g.db.execute("DELETE FROM todos WHERE id ='%s'" % id)
-    g.db.commit()
+
+    todo = Todos.query.filter_by(id=id).first()
+
+    description = todo.description
```
Comment (m-a-choquette) [inline alayatodo/views.py]: This will raise an `AttributeError` if the todo id doesn't exist.

File: tests/test_alayatodo.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+def test_login_logout(client):
+
+    ret = login(client, 'user1', 'user1')
+    assert b'user1 Logout' in ret.data
+
+    ret = logout(client)
+    assert b'You need to fork this repository.' in ret.data
+
+    ret = login(client, 'xyz', 123)
+    assert b'Unknown user. Please try again!' in ret.data
```
Comment (m-a-choquette) [inline tests/test_alayatodo.py]: This could be 3 tests. 
It's good to have 1 test testing 1 thing.

File: alayatodo/views.py [hunk 9]
```diff
@@ -44,39 +61,115 @@ def logout():
...
 def todos():
-    if not session.get('logged_in'):
-        return redirect('/login')
-    cur = g.db.execute("SELECT * FROM todos")
-    todos = cur.fetchall()
-    return render_template('todos.html', todos=todos)
+
+    page = request.args.get('page', 1, type=int)
+    user_id = session['user']['id']
+    todos = Todos.query.filter_by(user_id=user_id).paginate(page, app.config['TODOS_PER_PAGE'], False)
```
Comment (m-a-choquette) [inline alayatodo/views.py]: As per the pep8 guideline, this line is too long.

File: alayatodo/views.py [hunk 10]
```diff
@@ -44,39 +61,115 @@ def logout():
...
     return render_template('todo.html', todo=todo)
 
 
+@app.route('/todo/<id>/json', methods=['GET'])
+@todo_login
+def todo_json(id):
+
+    todo = Todos.query.filter_by(id=id).first()
+
+    response = {
```
Comment (m-a-choquette) [inline alayatodo/views.py]: This will raise `AttributeError`s if the todo doesn't exist.

File: tests/test_alayatodo.py [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+def test_todos(client):
+
+    login(client, 'user1', 'user1')
+
+    data = dict(description='')
+    ret = client.post('/todo/', data=data, follow_redirects=True)
+    assert b'Description required!' in ret.data
+
+    data = dict(description='another todo')
```
Comment (m-a-choquette) [inline tests/test_alayatodo.py]: Same  here.

**Comment by meistermuka** [comment]
Hello. Thank you for reviewing my pull request. I have addressed the changes as required.

**Pull request closed without merging**